Add messages for GetInstalledPackageSummaries and generate unimplemented client methods.

absoludity opened the pull request:
### Description of the change

Adds the messages from the design doc for the GetInstalledPackageSummaries call. I've noted clearly with comments on the PR where I differed and updated the messages (both here and on the design doc) from our last discussion about them.

I'm planning to do each addition as a separate PR so we can review the fields piecemeal here and if there's need for a larger discussion I'll organise a call and work through it together on the design doc. I expect that this will be necessary only for the later CreateInstalledPackage work, though perhaps also for the GetInstalledPackageDetail.

### Benefits

Enables implementation of the direct helm implementation of GetInstalledPackageSummaries.

### Applicable issues

<!-- Enter any applicable Issues here (You can reference an issue using #) -->
  - ref #3146

### Additional information

I'll comment inline on the couple of changes I made to the design doc while creating this PR.

absoludity commented on cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto:
```diff
@@ -108,6 +109,19 @@ message GetAvailablePackageVersionsRequest {
+message GetInstalledPackageSummariesRequest {
+  // The context (cluster/namespace) for the request.
+  Context context = 1;
+
```
We don't currently support filtering of installed packages at all, so I've removed the TODO from the design doc for here in the request (and in the doc).

absoludity commented on cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto:
```diff
@@ -306,6 +333,67 @@ message AvailablePackageDetail {
+  // Name
+  //
+  // A name given to the installation of the package.
+  string name = 2;
```
I added `name` here - the name of the installation summary (and in the design doc), because we'd forgotten to add it after we switched the `name` field in the installed package ref to a more opaque identifier a week or two ago (which we did to allow plugins to avoid ambiguous references based on the name).

gfichtenholt reviewed (approved):
I don't see anything glaringly wrong, so I approve. I am a little confused about the sequence of things though. We agreed on 3-4 APIs having to do with available packages. At the time we punted on other stuff  (install-related), as in didn't even discuss it, because we said (a) we were tired cuz (b) it takes a long time for all to agree on the perfect API and (c) the 3 -4 APIs we did agree on were sufficient to get some of the UX work started, like the catalog. So I thought that's what would come next. Instead you seemed to have gone ahead and added APIs we haven't discussed and agreed upon. Seems a little out of order, but I don't want to block you from doing this so go ahead

antgamdia reviewed (approved):
LGTM as well. I'm stating feeling that the current design doc is becoming quite messy and something not really useful.
I don't know if, perhaps, we can have a different approach for the upcoming operations: rather than discussing in gdocs, we could bring the discussion to a PR (that allows suggestions and conversations). 
Perhaps, opening a PR per operation (without any autogeneration, just the proto), could be an option. This way, we could resove the comments and add suggestions directly on the code (removing, thus, an extra step)
See what you think. Anyway, +1 in this PR, as Greg said

absoludity commented:
> I don't see anything glaringly wrong, so I approve. I am a little confused about the sequence of things though. We agreed on 3-4 APIs having to do with available packages.

Correct, and they are now implemented and landed for the direct helm plugin, so we can start there in the UI now.

I'm keen to be able to continue, so started looking at the next one which from memory we did actually get through (GetInstalledPackageSummaries) though we did indeed then realise we weren't going to get through the rest of the installed stuff, so we stopped and agreed to go ahead only with the subset - the available package API calls. So when I started looking at this one, it did indeed seem quite straight forward, so I thought a PR may suffice to identify anything for this one, though clearly the other calls (GetInstalledPackageDetail and especially the CreateInstalledPackage) will require higher bandwidth discussion.

> At the time we punted on other stuff (install-related), as in didn't even discuss it, because we said (a) we were tired cuz (b) it takes a long time for all to agree on the perfect API and (c) the 3 -4 APIs we did agree on were sufficient to get some of the UX work started, like the catalog. So I thought that's what would come next.

Yes, those API calls are implemented now for direct helm, which enables moving that part forward in the UX, which Antonio is looking at. While he's looking at the UI side, I plan to start moving the remaining APIs forward.

> Instead you seemed to have gone ahead and added APIs we haven't discussed and agreed upon. Seems a little out of order, but I don't want to block you from doing this so go ahead

I've not added them yet - this PR was to discuss the first ones before I'd go ahead and implement them in direct helm. I'm more than happy to have a call to go through the messages for GetInstalledPackageSummaries if it's needed. Let's organise one.

absoludity commented:
> LGTM as well. I'm stating feeling that the current design doc is becoming quite messy and something not really useful.

Please do try to help improve the doc. The only part I'm currently focusing on is the messages as it is what's blocking us. I think it is useful to first think about the bigger picture of all the messages, but it is a lot of context to hold. I'll go through the doc today and try to resolve any outstanding conversations. Or what in particular, about the messages for the API, are you finding messy if it's not the comments/conversations? (I agree other sections, some of which are us dumping thoughts for unresolved aspects specific to particular plugins, are tricky to clean).

> I don't know if, perhaps, we can have a different approach for the upcoming operations: rather than discussing in gdocs, we could bring the discussion to a PR (that allows suggestions and conversations).

That's exactly what I was trying to do here for `GetInstalledPackageSummaries`. Do you have any discussion points for this one?

> Perhaps, opening a PR per operation (without any autogeneration, just the proto), could be an option. This way, we could resove the comments and add suggestions directly on the code (removing, thus, an extra step)

This change here is just the proto and the related autogeneration (which is usually hidden in the diff anyway). I'm happy to remove the generated changes if that helps.

> See what you think. Anyway, +1 in this PR, as Greg said

Thanks.

gfichtenholt commented on cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto:
```diff
@@ -306,6 +333,67 @@ message AvailablePackageDetail {
+  // A name given to the installation of the package (eg. "my-postgresql-for-testing").
+  string name = 2;
+
+  // PkgVersionReference
```
PkgVersionReference => VersionReference?

absoludity replied:
This is the public name of the field on the go struct, not the type. It's confusing, but since the name in the protobuf below is `pkg_version_reference`, the corresponding Go struct is generated with go-naming conventions as `PkgVersionReference`. You can see that in the generated changes to cmd/kubeapps-apis/gen/core/packages/v1alpha1/packages.pb.go above.

Aside: So far, we have been documenting each field using the name of the field in the go struct, but as you'd expect, for the generated JS/Swagger, it's generated with JS naming conventions as `pkgVersionReference`, which you can see in the swagger changes above. The title for the swagger documentation is taken from this comment, so appears as `PkgVersionReference`. Perhaps we want to switch all our titles here to use the JS names for the swagger docs. Anyway, not something for here, imo.

gfichtenholt commented on cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto:
```diff
@@ -306,6 +333,67 @@ message AvailablePackageDetail {
+  // Installed package icon URL
+  //
+  // A url for an icon.
+  string icon_url = 5;
```
fluxv2 HelmRelease doesn't have a field for this. Would the plug-in leave that empty or have some logic to derive it from available package?

absoludity replied:
Yes the latter. The client (such as Kubeapps dashboard) wants to present to the user all the installed apps, including the icons (which come from the package, or in your specific case, the `HelmChart`). Similar for some other fields. Earlier on we looked at whether we wanted to instead include the available package nested in the installed package, but decided to use redundant fields on the installed package.

gfichtenholt commented on cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto:
```diff
@@ -306,6 +333,67 @@ message AvailablePackageDetail {
+  // ShortDescription
+  //
+  // A short description of the package (provided by the package)
+  string short_description = 7;
```
flux helmrelease doesn't provide these, so I'll need to pull these 2 off the corresponding helmchart? Just making sure

absoludity replied:
Yes.

gfichtenholt commented on cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto:
```diff
@@ -306,6 +333,67 @@ message AvailablePackageDetail {
+  // LatestPkgVersion
+  //
+  // The latest version available for this package, regardless of the pkg_version_reference.
+  string latest_pkg_version = 9;
```
Again, flux doesn't seem to provide these 2 in helmrelease.status field, which is where I'd expect to find them. Are these necessary to be able to display whether or not there is an upgrade available? If so, it *SEEMS* to me that the UX or core server can figure these out based on getAvailablePackageVersions. In fact, these two fields seem like they're about "available" more than "installed"

absoludity replied:
> Again, flux doesn't seem to provide these 2 in helmrelease.status field, which is where I'd expect to find them. Are these necessary to be able to display whether or not there is an upgrade available?

Yes, whether an upgrade is available (whether that's in the current version specification, ie. should auto-upgrade unless auto-reconcilliation is paused, or outside the current version specification, in which case a manual upgrade is required if desired.

> If so, it _SEEMS_ to me that the UX or core server can figure these out based on getAvailablePackageVersions. In fact, these two fields seem like they're about "available" more than "installed"

They're about both, at least, the first one is: latest matching pkg version is a function of the available versions that match the specific package version reference used for this installation. Yes, the UX can indeed figure these out based on fetching the available versions for each item in the paginated list and doing the semver work, but it can be quite complex, while (a) it's part of what a client needs to display a list of installed packages to the user, and (b) the data is all available to the plugin. If we did want to remove these from here, I think we'd at least need an API function to get latest matching pkg versions for an input list of (available_pkg_summary_ref, pkg_version_ref) tuples, so only one request is required to get the data for N installed packages that are being displayed.

Thoughts?

absoludity reviewed (commented):
Thanks Greg. Comments below.
